blockingconcurrentqueue.h

```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+  // Never allocates. Thread-safe.
+	inline bool wait_enqueue_timed(producer_token_t const& token, T const& item, std::int64_t timeout_usecs)
+  {
+		if (!sema_enqueue->wait(timeout_usecs)) {
+			return false;
+		}
+    while (!inner.try_enqueue(token, item)) {
+      continue;
+    }
+		sema_dequeue->signal();
+    return true;
+  }
+
+  // Blocks the current thread until either there is room for the element in the queue,
+  // or until the timeout requires.
+  // If enough room could be acquired, the element is enqueued (by copying it)
+  // using an explicit producer token and true is returned.
+  // If the timeout expired before the element could be enqueued, false is returned.
+  // Never allocates. Thread-safe.
+  template<typename Rep, typename Period>
+  inline bool wait_enqueue_timed(producer_token_t const& token, T const& item, std::chrono::duration<Rep, Period> const& timeout) {
+    return wait_enqueue_timed(token, item, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
+  }
+
+  // Blocks the current thread until either there is room for the element in the queue,
+  // or until the timeout requires.
+  // If enough room could be acquired, the element is enqueued (by moving it)
+  // using an explicit producer token and true is returned.
+  // If the timeout expired before the element could be enqueued, false is returned.
+  // Never allocates. Thread-safe.
+	inline bool wait_enqueue_timed(producer_token_t const& token, T &&item, std::int64_t timeout_usecs)
+  {
+		if (!sema_enqueue->wait(timeout_usecs)) {
+			return false;
+		}
+    while (!inner.try_enqueue(token, std::move(item))) {
+      continue;
+    }
+		sema_dequeue->signal();
+    return true;
+  }
+
+  // Blocks the current thread until either there is room for the element in the queue,
+  // or until the timeout requires.
+  // If enough room could be acquired, the element is enqueued (by moving it)
+  // using an explicit producer token and true is returned.
+  // If the timeout expired before the element could be enqueued, false is returned.
+  // Never allocates. Thread-safe.
+  template<typename Rep, typename Period>
+  inline bool wait_enqueue_timed(producer_token_t const& token, T &&item, std::chrono::duration<Rep, Period> const& timeout) {
+    return wait_enqueue_timed(token, item, std::chrono::duration_cast<std::chrono::microseconds>(timeout).count());
+  }
+
+  // Enqueues several items.
+  // Blocks if necessary before (and between) inserting items if not
+  // enough space is available for all the items.
+	// Note: Use std::make_move_iterator if the elements should be moved
+	// instead of copied.
+	// Does not allocate memory (except for one-time implicit producer).
+  // Thread-safe.
+	template<typename It>
+  inline void wait_enqueue_bulk(It itemFirst, size_t max)
+  {
+    size_t count = 0;
+    while (count < max) {
+		  size_t gained = (size_t)sema_enqueue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)(max - count));
```

Comment:
Not sure this is a good idea. It could lead to deadlock where two producers each acquire half the number of slots they need and are waiting on each other. Would be better to avoid a loop and use a single `waitManyExact` that waits for the entire count to become free.

Reply:
Hmm, I did think of that; the reason I ultimately decided to do it this because otherwise I could either only gurantee that *some* items are enqueued or I could never enqueue all of the elements passed. I'll go for the weaker variant, not guranteeing that all elements will be enqueued then…